### DISCUSSION TITLE
Duplicate logging with echo=True

### DISCUSSION THREAD

**Discussion by kshpytsya** [sqlalchemy/sqlalchemy#10302] [Usage Questions] [answered]
### Describe the bug

Due to what appears to be an incorrect check for presence of handlers here:

https://github.com/sqlalchemy/sqlalchemy/blob/b382bff6e3464f039db0fd1f2ce1b79038675e48/lib/sqlalchemy/log.py#L145

the `_add_default_handler` inserts an unnecessary extra handler.

I believe that a proper condition should use [hasHandlers](https://docs.python.org/3/library/logging.html#logging.Logger.hasHandlers) method instead.

### SQLAlchemy Version in Use

2.0.16

### DBAPI (i.e. the database driver)

irrelevant

### Database Vendor and Major Version

irrelevant

### Python Version

irrelevant

### Operating system

irrelevant

### To Reproduce

```python
irrelevant
```

### Error

irrelevant

**Comment by kshpytsya** [comment]
The following works as a workaround in app code:

```py
logging.getLogger("sqlalchemy.engine.Engine").addHandler(logging.NullHandler())
```

**Reply by simonvanlierde** [reply]
This sets a new NullHandler but doesn't remove the default handler, what works better for me is:
```python
logging.getLogger("sqlalchemy.engine.Engine").handlers = [logging.NullHandler()]
```

**Comment by CaselIT** [comment]
> To Reproduce

> irrelevant

To a discussion it goes

**Comment by CaselIT** [comment, marked as answer]
Note that the check seems correct. The idea is that echo is used when debugging or when an application is not using logging.

If you want to customize how the logging is used you can configure the level on the sqlalchemy logger externally

**Reply by kshpytsya** [reply]
Obviously I have misinterpreted the intent of "echo" parameter, which to my defense is not documented in `sqlalchemy.Engine.__init__` docstring. It was my understanding that it is a way to have Sqlalchemy log SQL statements. Now, diving deeper into docs I see the true intent of "echo" and a proper way of logging SQL statements. Thank you for your time!

**Reply by CaselIT** [reply]
Yes, it uses the standard logging module of python.

Echo is more "activate logging in all cases"

**Comment by zzzeek** [comment]
echo and logging overall with the engine is documented at https://docs.sqlalchemy.org/en/20/core/engines.html#configuring-logging

**Reply by CaselIT** [reply]
it has been like this since forever (since 98674959d25040d4d7ca3c95bbfb422cbb23c20e) but maybe the condition should evaluate to false then the level is notset? 

https://github.com/sqlalchemy/sqlalchemy/blob/c3097458d95448d11a6a6bf80add2f933c0c1907/lib/sqlalchemy/log.py#L140-L141

I think it would produce equivalent result, but would avoid adding two handlers is users us addHandler manually afterward

**Comment by mconigliaro** [comment]
I tried to set up [Rich Logging](https://rich.readthedocs.io/en/stable/logging.html) in my dev environment like this:

```python
logging.getLogger("sqlalchemy.engine").addHandler(RichHandler())
logging.getLogger("sqlalchemy.engine").setLevel(logging.INFO)
```

This resulted in duplicate log messages from `sqlalchemy.engine.Engine`, even with `echo=False` (maybe related to #10689?). I fixed it by adding this:

```python
logging.getLogger("sqlalchemy").handlers = [logging.NullHandler()]
```

**Reply by pot-code** [reply]
this is the answer

**Reply by CaselIT** [reply]
Note that this is just how python logging works. I'm not sure what sqlalchemy could do here.
Example:
```py
import logging

logging.basicConfig() # comment to avoid double lines

logging.getLogger("foo.bar").addHandler(logging.StreamHandler())
logging.getLogger("foo.bar").setLevel(logging.INFO)

logging.getLogger("foo.bar").info('abc')
```
this logs
```
abc
INFO:foo.bar:abc
```
